src/rust/lib/graph-editor/src/component/node/port.rs

```diff
@@ -0,0 +1,468 @@
+    type Shape = shape::Shape;
+    fn new(shape:&Self::Shape, _scene:&Scene,_shape_registry:&ShapeRegistry) -> Self {
+        let spec  = Specification::default();
+        shape.update_from_spec(&spec);
+        shape.is_inwards.set(1.0);
+
+        let bbox = Vector2::new(2.0 * shape.height.get(), 2.0 * shape.height.get());
+        shape.sprite.size().set(bbox);
+
+        Self {}
+    }
+}
+
+/// Shape view for output port.
+#[derive(Debug,Default,Clone,Copy)]
+pub struct OutputPortView {}
+impl ShapeViewDefinition for OutputPortView {
+    type Shape = shape::Shape;
+    fn new(shape:&Self::Shape, _scene:&Scene,_shape_registry:&ShapeRegistry) -> Self {
+        let spec  = Specification::default();
+        shape.update_from_spec(&spec);
+        shape.is_inwards.set(0.0);
+
+        let bbox = Vector2::new(2.0 * shape.height.get(), 2.0 * shape.height.get());
+        shape.sprite.size().set(bbox);
+
+        Self {}
+    }
+}
+
+/// Helper trait that describes a `ShapeViewDefinition` with a port shape.
+pub trait PortShapeViewDefinition = ShapeViewDefinition<Shape=shape::Shape>;
+
+/// Port definition.
+///
+/// A port must always be instantiated as an `InputPort` or an `OutputPort`. This determines its
+/// actual shape.
+///
+/// Example
+/// -------
+/// ```
+/// use graph_editor::component::node::port::InputPort;
+/// use graph_editor::component::node::port::OutputPort;
+///
+/// let input_port  = InputPort::default();
+/// let output_port = OutputPort::default();
+/// ```
+#[derive(Debug,CloneRef)]
+#[allow(missing_docs)]
+pub struct Port<T:PortShapeViewDefinition> {
+    pub data : Rc<PortData<T>>
+}
+
+impl<T:PortShapeViewDefinition> Clone for Port<T>{
+    fn clone(&self) -> Self {
+        let data = Rc::clone(&self.data);
+        Self {data}
+    }
+}
+
+/// Type that represents an input port.
+pub type InputPort = Port<InputPortView>;
+
+/// Type that represents an output port.
+pub type OutputPort = Port<OutputPortView>;
+
+/// Internal data of `Port.
+#[derive(Debug,Clone)]
+#[allow(missing_docs)]
+pub struct PortData<T:PortShapeViewDefinition> {
+        spec : RefCell<Specification>,
+    pub view : Rc<component::ShapeView<T>>
+}
+
+impl<T:PortShapeViewDefinition> Port<T> {
+
+    /// Internal constructor based on a given specification.
+    fn new(spec:Specification) -> Self {
+        let logger = Logger::new("node");
+        let view   = Rc::new(component::ShapeView::new(&logger));
+        let spec   = RefCell::new(spec);
+        let data   = Rc::new(PortData{spec,view});
+        Self {data} . init()
+    }
+
+    fn init(self) -> Self {
+        self.init_shape();
+        self.update_sprite_position();
+        self
+    }
+
+    /// Sets the port's position.
+    ///
+    /// Note this shadows the `set_position` function of `ObjectOps`.
+    pub fn set_position(&self, position:Angle<Degrees>) {
+        self.data.spec.borrow_mut().position = position;
+        self.update_sprite_position();
+    }
```

Comment:
`set_position` in DisplayObject` allwos setting 2D position. This is about setting rotation of the port. I would name it to `set_rotation`. Moreover, if we as user of this struct cannot use the properties of display object freely, then `Port` should not be a display object per-se, right? Of course its field `view` will be a `display::Object`, but we do not need to make `Port` one. Exposing all utils like `mod_position` or `mod_rotation` from display object doesn't make sense here. IMO you should remove `impl<'t,T:PortShapeViewDefinition> From<&'t Port<T>> for &'t display::object::Node` .